KAFKA-14590: Move DelegationTokenCommand to tools

tinaselenge opened the pull request:
Support delegationToken APIs in MockAdminClient.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tinaselenge committed:
KAFKA-14590: Move DelegationTokenCommand to tools
Support delegationToken APIs in MockAdminClient.

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+            System.err.println(e.getMessage());
+            return 1;
+        } catch (Throwable e) {
+            System.err.println("Error while executing delegation token command : " + e.getMessage());
```
I think we should only print the error message, because we know which command we are executing and there is a stack trace which follows.

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+            CommandLineUtils.maybePrintHelpOrVersion(opts, "This tool helps to create, renew, expire, or describe delegation tokens.");
+
+            // should have exactly one action
+            int numberOfAction = 0;
```
This block can be replaced by a one-liner and embedded in DelegationTokenCommandOptions.

```sh
int numberOfActions = Stream.of(opts.hasCreateOpt(), opts.hasRenewOpt(), opts.hasExpireOpt(), opts.hasDescribeOpt()).filter(b -> b).count();
```

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+
+            opts.checkArgs();
+
+            adminClient = createAdminClient(opts);
```
The Admin interface extends AutoClosable, so we can use try-with-resources.

fvaleri commented on build.gradle:
```diff
@@ -1763,6 +1763,7 @@ project(':tools') {
     implementation libs.jacksonJDK8Datatypes
     implementation libs.slf4jApi
     implementation libs.log4j
+    implementation libs.joptSimple
```
This change is already merged, so you can remove it and rebase.

clolov replied:
If you rebase on trunk there won't be a need for this anymore 😊

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    static class DelegationTokenCommandOptions extends CommandDefaultOptions {
+        public final ArgumentAcceptingOptionSpec<String> bootstrapServerOpt;
```
Unless required, I would suggest to always use OptionSpec interface.

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+        List<KafkaPrincipal> principals = new ArrayList<>();
+
+        if (opts.options.has(principalOptionSpec)) {
+            for (Object e : opts.options.valuesOf(principalOptionSpec))
```
Why are you using Object instead of String here?

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+
+    public static DelegationToken createToken(Admin adminClient, DelegationTokenCommandOptions opts) throws ExecutionException, InterruptedException {
+        List<KafkaPrincipal> renewerPrincipals = getPrincipals(opts, opts.renewPrincipalsOpt);
+        Long maxLifeTimeMs = opts.options.valueOf(opts.maxLifeTimeOpt);
```
I would embed all `opts.options...` in DelegationTokenCommandOptions.

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+                    dateFormat.format(tokenInfo.issueTimestamp()),
+                    dateFormat.format(tokenInfo.expiryTimestamp()),
+                    dateFormat.format(tokenInfo.maxTimestamp()));
+            System.out.println();
```
I think we don't need this one.

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+        if (ownerPrincipals.isEmpty()) {
+            System.out.println("Calling describe token operation for current user.");
+        } else {
+            System.out.println("Calling describe token operation for owners :" + ownerPrincipals);
```
```suggestion
            System.out.printf("Calling describe token operation for owners: %s%n", ownerPrincipals);
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+                    .withRequiredArg()
+                    .ofType(String.class);
+
+            this.createOpt = parser.accepts("create", "Create a new delegation token. Use --renewer-principal option to pass renewers principals.");
```
```suggestion
            this.createOpt = parser.accepts("create", "Create a new delegation token. Use --renewer-principal option to pass renewer principals.");
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+                    .ofType(String.class);
+
+            this.maxLifeTimeOpt = parser.accepts("max-life-time-period", "Max life period for the token in milliseconds. If the value is -1," +
+                            " then token max life time will default to a server side config the value (delegation.token.max.lifetime.ms).")
```
```suggestion
                            " then token max life time will default to the server side config value of (delegation.token.max.lifetime.ms).")
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+                    .ofType(Long.class);
+
+            this.renewTimePeriodOpt = parser.accepts("renew-time-period", "Renew time period in milliseconds. If the value is -1, then the" +
+                            " renew time period will default to a server side config the value (delegation.token.expiry.time.ms).")
```
```suggestion
                            " renew time period will default to the server side config value of (delegation.token.expiry.time.ms).")
```

fvaleri reviewed (changes requested):
Hi @tinaselenge, thanks for working on this.

I left some comments and there are some checkstyle issues.

clolov commented on tools/src/test/java/org/apache/kafka/tools/DelegationTokenCommandTest.java:
```diff
@@ -0,0 +1,105 @@
+    private DelegationTokenCommand.DelegationTokenCommandOptions getDescribeOpts(String owner) {
+        String[] args;
+        if (!owner.equals("")) {
+            args = new String[] {"--bootstrap-server", "localhost:9092", "--command-config", "testfile", "--describe","--owner-principal", owner};
```
```suggestion
            args = new String[] {"--bootstrap-server", "localhost:9092", "--command-config", "testfile", "--describe", "--owner-principal", owner};
```

clolov commented on tools/src/test/java/org/apache/kafka/tools/DelegationTokenCommandTest.java:
```diff
@@ -0,0 +1,105 @@
+        } else {
+            args = new String[] {"--bootstrap-server", "localhost:9092", "--command-config", "testfile", "--describe"};
+        }
+        return new DelegationTokenCommand.DelegationTokenCommandOptions(args);
```
Maybe the following will save us a few repetitive strings?
```suggestion
        List<String> args = new ArrayList() {{
          .add("--bootstrap-server");
          .add("localhost:9092");
          .add("--command-config");
          .add("testfile");
          .add("--describe")
        }}
        if (!owner.equals("")) {
          args.add("--owner-principal");
          args.add(owner);
        }
        return new DelegationTokenCommand.DelegationTokenCommandOptions(args);
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+            this.renewOpt = parser.accepts("renew", "Renew delegation token. Use --renew-time-period option to set renew time period.");
+            this.expiryOpt = parser.accepts("expire", "Expire delegation token. Use --expiry-time-period option to expire the token.");
+            this.describeOpt = parser.accepts("describe", "Describe delegation tokens for the given principals. Use --owner-principal to pass owner/renewer principals." +
+                    " If --owner-principal option is not supplied, all the user owned tokens and tokens where user have Describe permission will be returned.");
```
```suggestion
                    " If --owner-principal option is not supplied, all the user-owned tokens and tokens where the user has Describe permissions will be returned.");
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+            this.describeOpt = parser.accepts("describe", "Describe delegation tokens for the given principals. Use --owner-principal to pass owner/renewer principals." +
+                    " If --owner-principal option is not supplied, all the user owned tokens and tokens where user have Describe permission will be returned.");
+
+            this.ownerPrincipalsOpt = parser.accepts("owner-principal", "owner is a kafka principal. It is should be in principalType:name format.")
```
```suggestion
            this.ownerPrincipalsOpt = parser.accepts("owner-principal", "owner is a Kafka principal. They should be in principalType:name format.")
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+                    .withOptionalArg()
+                    .ofType(String.class);
+
+            this.renewPrincipalsOpt = parser.accepts("renewer-principal", "renewer is a kafka principal. It is should be in principalType:name format.")
```
```suggestion
            this.renewPrincipalsOpt = parser.accepts("renewer-principal", "renewer is a Kafka principal. They should be in principalType:name format.")
```

clolov commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -596,22 +604,89 @@ synchronized public DeleteRecordsResult deleteRecords(Map<TopicPartition, Record
+        }
+
+        String tokenId = Uuid.randomUuid().toString();
+        TokenInformation tokenInfo = new TokenInformation(tokenId,options.renewers().get(0), options.renewers(),System.currentTimeMillis(), options.maxlifeTimeMs(), -1);
```
```suggestion
        TokenInformation tokenInfo = new TokenInformation(tokenId, options.renewers().get(0), options.renewers(), System.currentTimeMillis(), options.maxlifeTimeMs(), -1);
```

clolov commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -596,22 +604,89 @@ synchronized public DeleteRecordsResult deleteRecords(Map<TopicPartition, Record
+
+        String tokenId = Uuid.randomUuid().toString();
+        TokenInformation tokenInfo = new TokenInformation(tokenId,options.renewers().get(0), options.renewers(),System.currentTimeMillis(), options.maxlifeTimeMs(), -1);
+        DelegationToken token = new DelegationToken(tokenInfo,tokenId.getBytes());
```
```suggestion
        DelegationToken token = new DelegationToken(tokenInfo, tokenId.getBytes());
```

clolov commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -94,6 +100,8 @@ public class MockAdminClient extends AdminClient {
 
     private Map<MetricName, Metric> mockMetrics = new HashMap<>();
 
+    private final List<DelegationToken> alltokens = new ArrayList<>();
```
```suggestion
    private final List<DelegationToken> allTokens = new ArrayList<>();
```

clolov commented on tools/src/main/java/org/apache/kafka/tools/DelegationTokenCommand.java:
```diff
@@ -0,0 +1,306 @@
+
+        } finally {
+            if (adminClient != null)
+                adminClient.close();
```
Since Admin implements AutoCloseable could you use https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html?

tinaselenge committed:
Address review comments

fvaleri reviewed (approved):
LGTM.

There is still one checkstyle issue that needs to be fixed.

tinaselenge committed:
Address checkstyle issue

tinaselenge commented:
Thank you @fvaleri  and @clolov for reviewing the PR.

showuon reviewed (commented):
@tinaselenge , thanks for the PR. Some questions:
1. This PR creates a new `DelegationTokenCommand` class, but there's no old `DelegationTokenCommand` class removal. Why is that?
2. The original `DelegationTokenCommandTest` is an integration test, but now we changed to unit test by mockAdminClient, why do we change that?

Thanks.

tinaselenge committed:
Remove the existing tool in Scala

tinaselenge commented:
> @tinaselenge , thanks for the PR. Some questions:
> 
> 1. This PR creates a new `DelegationTokenCommand` class, but there's no old `DelegationTokenCommand` class removal. Why is that?
> 2. The original `DelegationTokenCommandTest` is an integration test, but now we changed to unit test by mockAdminClient, why do we change that?
> 
> Thanks.

Hi @showuon 

1. I have removed the existing Scala class and its test.

2. I thought it's good enough to test it using the mock as it's not really doing anything specific with the cluster. I understand that changes the test behaviour. If you think we should test the tool against an integration test cluster, I'm happy to change it back. Please let me know.

showuon commented:
> > @tinaselenge , thanks for the PR. Some questions:
> > 
> > 1. This PR creates a new `DelegationTokenCommand` class, but there's no old `DelegationTokenCommand` class removal. Why is that?
> > 2. The original `DelegationTokenCommandTest` is an integration test, but now we changed to unit test by mockAdminClient, why do we change that?
> > 
> > Thanks.
> 
> Hi @showuon
> 
>     1. I have removed the existing Scala class and its test.
> 
>     2. I thought it's good enough to test it using the mock as it's not really doing anything specific with the cluster. I understand that changes the test behaviour. If you think we should test the tool against an integration test cluster, I'm happy to change it back. Please let me know.

For (1), thanks for the update
For (2), yes, I think the integration test is important, especially after KRaft starting to support delegation token, we should rely on integration tests to make sure everything works fine on both ZK and KRaft mode. But the mock client implementation are also great to be added. Those could also be kept. 

Thank you.

showuon commented:
@tinaselenge ,  for (2), it looks complicated since under the tools module, there's no `BaseRequestTest` or `SaslSetup` to help setup the delegation token integration tests. Since it might make this PR much more complicated, I think we can create another JIRA ticket to track this integration task, so that it won't block the current task. WDYT?

tinaselenge commented:
@showuon I agree. I will create a Jira issue to create an integration test for this.

tinaselenge commented:
Created a Jira issue for adding integration test for this tool later on: https://issues.apache.org/jira/browse/KAFKA-14763

showuon reviewed (approved):
LGTM!

showuon commented:
Failed tests are unrelated
```
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testUnregisterBroker()
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest.testBrokerCoordinator
    Build / JDK 8 and Scala 2.12 / kafka.server.KafkaServerKRaftRegistrationTest.[1] Type=ZK, Name=testRegisterZkBrokerInKraft, MetadataVersion=3.4-IV0, Security=PLAINTEXT
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.internals.EagerConsumerCoordinatorTest.testPrepareJoinAndRejoinAfterFailedRebalance()
    Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testTrustStoreAlter(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testTrustStoreAlter(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.processor.internals.StreamsAssignmentScaleTest.testHighAvailabilityTaskAssignorManyStandbys
```
